File: src/SimpleLevelEditor.Formats.Types/Level/WorldObject.fs
namespace SimpleLevelEditor.Formats.Types.Level

open System
open System.Numerics

type WorldObject =
    { Id: int32 // The id is only used to keep track of the object in the editor.
      mutable Mesh: string // TODO: Rename to ModelPath.
      mutable Texture: string // TODO: Remove.
      mutable Scale: Vector3
      mutable Rotation: Vector3
      mutable Position: Vector3
      mutable Flags: string list }

    member this.DeepCopy() =
        { this with
            Flags = this.Flags |> List.map id }

    member this.CloneAndPlaceAtPosition(entityId: int32, position: Vector3) =
        { Id = entityId
          Mesh = this.Mesh
          Texture = this.Texture
          Scale = this.Scale
          Rotation = this.Rotation
          Position = position
          Flags = this.Flags |> List.map id }

    // TODO: Test this member.
    member this.ChangeFlagAtIndex(flagIndex: int32, flag: string) =
        if flagIndex >= 0 && flagIndex < this.Flags.Length then
            this.Flags <- this.Flags |> List.mapi (fun i v -> if i = flagIndex then flag else v)

    // TODO: Test this member.
    member this.AddFlag(flag: string) =
        this.Flags <- this.Flags @ [flag]

    // TODO: Test this member.
    member this.RemoveFlagAtIndex(flagIndex: int32) =
        if flagIndex >= 0 && flagIndex < this.Flags.Length then
            this.Flags <- this.Flags |> List.mapi (fun i v -> if i = flagIndex then null else v) |> List.filter (fun v -> v <> null)

    static member CreateDefault() =
        { Id = 0
          Mesh = String.Empty
          Texture = String.Empty
          Scale = Vector3.One
          Rotation = Vector3.Zero
          Position = Vector3.Zero
          Flags = [] }

File: src/SimpleLevelEditor.Formats.Level/WorldObject.cs
using System.Numerics;

namespace SimpleLevelEditor.Formats.Level;

public sealed record WorldObject
{
	public required int Id { get; init; }

	public required string ModelPath { get; set; }

	public required Vector3 Scale { get; set; }

	public required Vector3 Rotation { get; set; }

	public required Vector3 Position { get; set; }

	public required List<string> Flags { get; set; }

	public WorldObject DeepCopy()
	{
		return this with
		{
			Flags = Flags.ToList(),
		};
	}

	public WorldObject CloneAndPlaceAtPosition(int objectId, Vector3 position)
	{
		return this with
		{
			Id = objectId,
			Position = position,
			Flags = Flags.ToList(),
		};
	}

	public void ChangeFlagAtIndex(int index, string flag)
	{
		if (index < 0 || index >= Flags.Count)
			return;

		Flags[index] = flag;
	}

	public void AddFlag(string flag)
	{
		Flags.Add(flag);
	}

	public void RemoveFlagAtIndex(int index)
	{
		if (index < 0 || index >= Flags.Count)
			return;

		Flags.RemoveAt(index);
	}

	public static WorldObject CreateDefault()
	{
		return new WorldObject
		{
			Id = 0,
			ModelPath = string.Empty,
			Scale = Vector3.One,
			Rotation = Vector3.Zero,
			Position = Vector3.Zero,
			Flags = [],
		};
	}
}
